treeherder/perf/data_cycling/removal_strategies.py

```diff
@@ -0,0 +1,364 @@
+            .exclude(repository_id__in=self.relevant_repositories)
+            .order_by('-id')[0]
+            .id
+        )
+        older_perf_data_rows = (
+            self._manager.filter(
+                push_timestamp__lte=self._max_timestamp, id__lte=max_id_of_non_expired_row
+            )
+            .exclude(repository_id__in=self.relevant_repositories)
+            .order_by('id')[: self._chunk_size]
+        )
+        return len(older_perf_data_rows) or self._chunk_size
+
+
+class StalledDataRemoval(RemovalStrategy):
+    """
+    Removes `performance_datum` rows from `performance_signature`s
+    that haven't been updated in the last 4 months.
+    """
+
+    @property
+    def CYCLE_INTERVAL(self) -> int:
+        # WARNING!! Don't override this without proper approval!
+        return 120  # days                                     #
+        ########################################################
+
+    def __init__(self, chunk_size: int, days: int = None):
+        super().__init__(chunk_size, days=days)
+
+        self._target_signature = None
+        self._removable_signatures = None
+
+    @property
+    def target_signature(self) -> PerformanceSignature:
+        try:
+            if self._target_signature is None:
+                self._target_signature = self.removable_signatures.pop()
+        except IndexError:
+            msg = 'No stalled signature found.'
+            logger.warning(msg)  # no stalled data is not normal
+            raise LookupError(msg)
+        return self._target_signature
+
+    @property
+    def removable_signatures(self) -> List[PerformanceSignature]:
+        if self._removable_signatures is None:
+            self._removable_signatures = list(
+                PerformanceSignature.objects.filter(last_updated__lte=self._max_timestamp).order_by(
+                    'last_updated'
+                )
+            )
+        return self._removable_signatures
+
+    def remove(self, using: CursorWrapper):
+        while True:
+            try:
+                self.__attempt_remove(using)
+
+                deleted_rows = using.rowcount
+                if deleted_rows > 0:
+                    break  # deletion was successful
+
+                self.__lookup_new_signature()  # to remove data from
+            except LookupError as ex:
+                logger.debug(
+                    f'Could not target any (new) stalled signature to delete data from. {ex}'
+                )
+                break
+
+    @property
+    def max_timestamp(self) -> datetime:
+        return self._max_timestamp
+
+    @property
+    def name(self) -> str:
+        return 'stalled data removal strategy'
+
+    def __attempt_remove(self, using: CursorWrapper):
+        using.execute(
+            '''
+                DELETE FROM `performance_datum`
+                WHERE repository_id = %s AND signature_id = %s AND push_timestamp <= %s
+                LIMIT %s
```

Comment:
I'm curious why this is using raw sql instead of the queryset API.

Reply:
It's using raw sql because of an old Django limitation. The queryset API isn't able build a delete statement with a `LIMIT` in it, even though the underlying MySQL database is capable of that. And this `LIMIT` is essential, as it's the only optimal way we found to remove data.

In the past, we tried to delete chunks of data using the queryset API, but 90% of the time it timed out. That's why we needed to go deeper into the SQL of it.

Reply:
Makes sense - thanks for explaining.

Reply:
Sure thing. I'm going to inline the comments to each of the raw SQL sections.